### PR TITLE
Sub-grid parallelism support for BN and Topology-aware designs

### PR DESCRIPTION
**Pull request by aj-prime**
This PR adds sub-grid parallelism support for Batch Normalization and topology-aware designs along with AL communication primitives for communication in sub-grid parallelism.

### CODE REVIEW

File: src/execution_algorithms/kfac/kfac_util.cpp [hunk 1]
```diff
@@ -430,6 +430,680 @@ void allgather_blocks(
...
+void TranslateBetweenGridsVC
+(El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,El::Device::CPU> const& A,
+  El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,El::Device::CPU>& B){
+  LBANN_ERROR("TranslateBetweenGridsVC function is not implemented for CPUs");
+}
+
+template<typename T, El::Device Device>
+void TranslateBetweenGridsVC
+(El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device> const& A,
+  El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device>& B)
```
Comment (timmoon10) [inline src/execution_algorithms/kfac/kfac_util.cpp:502]: These template definitions should be reversed. The valid implementation is narrowly targeted toward a specific data type and device, so it should be the override instead of the default. This is also more amenable to CPU builds.
```suggestion
namespace {
#ifdef LBANN_HAS_GPU
using KFACDevice = El::Device::GPU;
#else
using KFACDevice = El::Device::CPU;
#endif // LBANN_HAS_GPU
} // namespace <anon>

template<typename T, El::Device Device>
void TranslateBetweenGridsVC
(El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device> const& A,
  El::DistMatrix<T,El::STAR,El::VC,El::ELEMENT,Device>& B){
  LBANN_ERROR("TranslateBetweenGridsVC function is not implemented for this configuration");
}

template<>
void TranslateBetweenGridsVC
(El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,KFACDevice> const& A,
  El::DistMatrix<DataType,El::STAR,El::VC,El::ELEMENT,KFACDevice>& B)
```

File: src/execution_algorithms/kfac/kfac_block_fc_conv.cpp [hunk 2]
```diff
@@ -369,6 +379,293 @@ void kfac_block_fc_conv<Device>::update_kronecker_inverse(
...
+void kfac_block_fc_conv<Device>::end_communication_forward_end(
+    lbann_comm* comm) {
+  if((comm->get_grid_type() == GridType::SECONDARY_GRID or comm->get_grid_type() == GridType::PRIMARY_GRID) 
+      and comm->enable_subgrid_async_communication()
+      and comm->get_KFAC_subgrid_create_two_models()==false){
+    auto primary_grid_ranks = comm->get_primary_grid_ranks();
+    auto secondary_grid_ranks = comm->get_secondary_grid_ranks();
+
+    for(auto& req:this->m_requests_forward_end){
+      ::Al::Wait<::Al::NCCLBackend>(req);
```
Comment (timmoon10) [inline src/execution_algorithms/kfac/kfac_block_fc_conv.cpp:457]: I see build errors on Catalyst since `Al::NCCLBackend` isn't defined. They should all be replaced with `BackendT`:
```suggestion
      ::Al::Wait<BackendT>(req);
```

**Review by timmoon10 — DISMISSED** [review]
Overall looks correct, although I have comments on style and code structure.

[Here are the Bamboo results](https://lc.llnl.gov/bamboo/browse/LBANN-TIM409-1). There are no new failures on Lassen, but many compile errors on Catalyst (skimming the logs, it looks like undefined references to `Al::NCCLBackend`). We should also make sure this doesn't disrupt the Bamboo on Corona, once that is working again with the nightly build.

File: include/lbann/execution_algorithms/kfac/kfac_block.hpp [hunk 3]
```diff
@@ -48,17 +60,30 @@ class kfac_block {
...
     m_has_kronecker_inverse = false;
   }
   virtual ~kfac_block() = default;
 
   virtual void on_forward_prop_end(lbann_comm* comm) {}
 
+  /** @brief Get local Memory Consumption. */
+  virtual int get_local_memory_consumption() {
+    LBANN_ERROR("this function should be called via a sub-class.");
+  }
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block.hpp:85]:
```suggestion
  virtual int get_local_memory_consumption() = 0;
```
Is there a reason these aren't left pure virtual??

File: include/lbann/execution_algorithms/kfac/kfac_block.hpp [hunk 4]
```diff
@@ -114,6 +140,24 @@ class kfac_block {
...
+  virtual void start_communication_backward_end(
+      lbann_comm* comm){
+    LBANN_ERROR("this function should be called via a sub-class.");
+  }
+  virtual void end_communication_backward_end(
+      lbann_comm* comm){
+    LBANN_ERROR("this function should be called via a sub-class.");
+  }
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block.hpp:158]:
```suggestion
  virtual void start_communication_forward_end(lbann_comm* comm) = 0;
  virtual void end_communication_forward_end(lbann_comm* comm) = 0;
  virtual void start_communication_backward_end(lbann_comm* comm) = 0;
  virtual void end_communication_backward_end(lbann_comm* comm) = 0;
```
Is there a reason these aren't left pure virtual??

File: include/lbann/execution_algorithms/kfac/kfac_block.hpp [hunk 5]
```diff
@@ -204,6 +264,12 @@ class kfac_block {
       TODO: Remove this. */
   const size_t m_layer_id;
 
+  /** @brief Enable copying of errors to enhance async communication. */
+  bool m_enable_copy_errors;
+
+  /** @brief Enable copying of activations to enhance async communication. */
+  bool m_enable_copy_activations;
+
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block.hpp:272]: We don't really optimize our packing, but putting two `bool`s between a `size_t` and an `int` effectively guarantees 2 extra pad bytes/object... I would move these with `m_has_kronecker_inverse` below, or just move all 3 `bool`s to the end of the struct.

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+#include "lbann/execution_algorithms/kfac/kfac_block.hpp"
+
+namespace lbann {
+
+
+/** An FC/conv building block for K-FAC.
+ * TODO: Split into kfac_block_fc and kfac_block_conv.
+ */
+template <El::Device Device>
+class kfac_block_channelwise_fc: public kfac_block<Device> {
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:39]:
```suggestion
class kfac_block_channelwise_fc final : public kfac_block<Device> {
```
This class should be `final`, right?

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+                     const int output_size)
+      : kfac_block<Device>(layer, context, layer_id, inverse_proc_rank, enable_copy_errors, enable_copy_activations, input_size, output_size),
+        m_has_bias(layer->num_weights() > 1) {
+
+  }
+
+  kfac_block_channelwise_fc(const kfac_block_channelwise_fc&) = default;
+  kfac_block_channelwise_fc& operator=(const kfac_block_channelwise_fc&) = default;
+
+  int get_local_memory_consumption() override {
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:60]:
```suggestion
  int get_local_memory_consumption() final {
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+  void resize_inverse_matrices_size(El::Matrix<double, El::Device::CPU>& inverse_matrices_size, int block_number) override;
+
+
+  int set_inverse_matrices(
+      El::Matrix<DataType, Device>& workspace,
+      int offset,
+      lbann_comm *comm) override;
+
+  std::string get_info() const override {
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:141]:
```suggestion
  std::string get_info() const final {
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+  std::vector<int> get_inverse_matrices_size_vector(lbann_comm *comm) override;
+
+  void resize_inverse_matrices_size(El::Matrix<double, El::Device::CPU>& inverse_matrices_size, int block_number) override;
+
+
+  int set_inverse_matrices(
+      El::Matrix<DataType, Device>& workspace,
+      int offset,
+      lbann_comm *comm) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:139]:
```suggestion
      lbann_comm *comm) final;
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+  int get_inverse_matrices(
+      El::Matrix<DataType, Device>& output,
+      int offset) override;
+
+  int get_inverse_matrices_size(lbann_comm *comm) override;
+
+  std::vector<int> get_inverse_matrices_size_vector(lbann_comm *comm) override;
+
+  void resize_inverse_matrices_size(El::Matrix<double, El::Device::CPU>& inverse_matrices_size, int block_number) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:133]:
```suggestion
  int get_inverse_matrices_size(lbann_comm *comm) final;

  std::vector<int> get_inverse_matrices_size_vector(lbann_comm *comm) final;

  void resize_inverse_matrices_size(El::Matrix<double, El::Device::CPU>& inverse_matrices_size, int block_number) final;
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+      int num_weights) override;
+
+  void start_communication_forward_end(
+      lbann_comm* comm) override;
+  void end_communication_forward_end(
+      lbann_comm* comm) override;
+  void start_communication_backward_end(
+      lbann_comm* comm) override;
+  void end_communication_backward_end(
+      lbann_comm* comm) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:120]:
```suggestion
  void start_communication_forward_end(lbann_comm* comm) final;
  void end_communication_forward_end(lbann_comm* comm) final;
  void start_communication_backward_end(lbann_comm* comm) final;
  void end_communication_backward_end(lbann_comm* comm) final;
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+
+  const std::vector<El::AbstractMatrix<DataType>*>
+  get_preconditioned_grad_buffers() override;
+
+  int get_inverse_matrices(
+      El::Matrix<DataType, Device>& output,
+      int offset) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:127]:
```suggestion
  get_preconditioned_grad_buffers() final;

  int get_inverse_matrices(
      El::Matrix<DataType, Device>& output,
      int offset) final;
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    return total_size;
+  }
+
+  void compute_local_kronecker_factors(
+      lbann_comm* comm,
+      bool print_matrix,
+      bool print_matrix_summary) override;
+
+  const std::vector<El::AbstractMatrix<DataType>*>
+  get_local_kronecker_buffers() override {
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:78]:
```suggestion
      bool print_matrix_summary) final;

  const std::vector<El::AbstractMatrix<DataType>*>
  get_local_kronecker_buffers() final {
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+    std::vector<El::AbstractMatrix<DataType>*> ret =
+        {&m_kronecker_factor_buf_A, &m_kronecker_factor_buf_G};
+    return ret;
+  }
+
+  void update_kronecker_average(
+      lbann_comm* comm,
+      DataType kronecker_decay,
+      bool print_matrix,
+      bool print_matrix_summary) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:88]:
```suggestion
      bool print_matrix_summary) final;
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+
+  void update_kronecker_inverse(
+      lbann_comm* comm,
+      bool use_pi,
+      DataType damping_act, DataType damping_err,
+      DataType learning_rate_factor,
+      bool use_eigen_decomposition,
+      bool print_matrix,
+      bool print_matrix_summary,
+      bool print_time) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:98]:
```suggestion
      bool print_time) final;
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+  void compute_preconditioned_gradients(
+      lbann_comm* comm,
+      DataType learning_rate_factor,
+      bool print_matrix,
+      bool print_matrix_summary,
+      bool print_time) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:105]:
```suggestion
      bool print_time) final;
```

File: include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+
+  void initialize_activations_and_errors(
+      lbann_comm* comm,
+      int num_local_activations,
+      int num_local_errors,
+      int num_weights) override;
```
Comment (benson31) [inline include/lbann/execution_algorithms/kfac/kfac_block_channelwise_fc.hpp:111]:
```suggestion
      int num_weights) final;
```

File: src/execution_algorithms/kfac.cpp [hunk 18]
```diff
@@ -87,9 +95,68 @@ KFAC::KFAC(
...
-    m_compute_interval{compute_interval}
+    m_compute_interval{compute_interval},
+    m_distribute_precondition_compute{distribute_precondition_compute},
+    m_enable_copy_errors{enable_copy_errors},
+    m_enable_copy_activations{enable_copy_activations},
+    m_use_eigen_decomposition{use_eigen_decomposition},
+    m_use_KFAC_epoch{std::move(kfac_use_interval)}
 {}
 
+// KFAC::KFAC(KFAC const& other)
```
Comment (benson31) [inline src/execution_algorithms/kfac.cpp:106]: If this class is not Copyable, that's fine, but can we just remove the dead code?

File: src/execution_algorithms/kfac.cpp [hunk 19]
```diff
@@ -162,9 +233,10 @@ void KFAC::train(
       dc.reset_mode(sgd_context);
       do_epoch_begin_cbs(model);
       is_start_of_epoch = false;
+      // std::cout<<"Transfer Size:"<<m_global_inverse_buffer_size<<" Inverse time:" << m_time_span_inverse_comm<<" Inverse comm time:"<<m_time_span_inverse_send_recv<<"\n";
```
Comment (benson31) [inline src/execution_algorithms/kfac.cpp:236]: Remove debugging code?

File: src/execution_algorithms/kfac.cpp [hunk 20]
```diff
@@ -162,9 +233,10 @@ void KFAC::train(
...
       //sync weights if we have a separate model for primary and secondary grid
-      if(comm.get_KFAC_subgrid_create_two_models() and comm.get_grid_type()!=GridType::NO_GRID)
-        sync_weights_model(model, model.get_comm());
+      // if(comm.get_KFAC_subgrid_create_two_models() and comm.get_grid_type()!=GridType::NO_GRID)
+      // start_sync_weights_async(model, model.get_comm());
```
Comment (benson31) [inline src/execution_algorithms/kfac.cpp:239]: Remove dead code?

File: src/execution_algorithms/kfac.cpp [hunk 21]
```diff
@@ -173,6 +245,25 @@ void KFAC::train(
...
 
+      //profiling code
+      // std::cout<<"Comm rank:"<<comm.get_rank_in_world()<<" Primary Grid:"<< (comm.get_grid_type() == GridType::PRIMARY_GRID) <<" Inverse comm:"<<m_time_span_inverse_comm \
+      // <<" backward:"<<m_time_span_backward_comm<<" "<<m_time_span_backward_comm_end \
+      // <<" forward:"<<m_time_span_forward_comm <<" "<<m_time_span_forward_comm_end \
+      // <<" InvPre:"<<m_time_span_precond_comm
+      // <<" Forward:"<<m_time_forward_pass
+      // <<" backward:"<<m_time_backward_pass
+      // <<" KFAC:"<<m_time_kfac
+      // <<" Size of"<<sizeof(DataType)<<"\n";
```
Comment (benson31) [inline src/execution_algorithms/kfac.cpp:256]: Remove?

**Review by benson31 — APPROVED** [review]
There are some rough edges here. I have a weak preference to smooth them out now, but I won't hold up the merge for it.

In general, there seems to be a lot of new commented-out code. I'd prefer to see that either gone or factored into conditional code (e.g., the benchmarking code).

File: include/lbann/comm.hpp [hunk 22]
```diff
@@ -179,7 +179,10 @@ class lbann_comm
   void split_trainers(int procs_per_trainer=-1, int trainer_grid_height=-1);
 
   /** Split the commicator for the given trainer into primary and seconday*/
-  void split_trainer_grid(int num_process_primary_grid=0, bool create_two_models=false);
+  void split_trainer_grid(int num_process_primary_grid=0,
+                          bool create_two_models=false,
+                          bool enable_async_comm=false,
+                          bool enable_topo_aware=false);
```
[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/comm.hpp:197]: Done.

**Review by bvanessen — APPROVED** [review]

**Review by benson31 — APPROVED** [review]
LGTM

**Review by bvanessen — APPROVED** [review]
